internal/monitors/kubernetes/utils/servicepodcache.go

```diff
@@ -0,0 +1,239 @@
+				if _, exists := psc.podSvcUIDCache[podUID]; !exists {
+					psc.podSvcUIDCache[podUID] = make(map[types.UID]bool)
+
+				}
+				if _, exists := psc.svcPodUIDCache[svc.UID]; !exists {
+					psc.svcPodUIDCache[svc.UID] = make(map[types.UID]bool)
+				}
+				pods = append(pods, podUID)
+				psc.podSvcUIDCache[podUID][svc.UID] = true
+				psc.svcPodUIDCache[svc.UID][podUID] = true
+			}
+		}
+	}
+	return pods
+}
+
+// refreshCacheByPod should be called when a pod is added
+// or updated and the pod:service mappings need to be refreshed.
+// This function loops through all services in the cache and checks if
+// any match the given pod.
+func (psc *PodServiceCache) refreshCacheByPod(pod *v1.Pod) []types.UID {
+	var services []types.UID
+	if labelSet, exists := psc.podUIDLabelCache[pod.UID]; exists {
+		for svcUID, selector := range psc.svcUIDSelectorCache {
+			if selector.Matches(labelSet) &&
+				psc.svcUIDNamespaceCache[svcUID] == pod.Namespace {
+				if _, exists := psc.podSvcUIDCache[pod.UID]; !exists {
+					psc.podSvcUIDCache[pod.UID] = make(map[types.UID]bool)
+
+				}
+				if _, exists := psc.svcPodUIDCache[svcUID]; !exists {
+					psc.svcPodUIDCache[svcUID] = make(map[types.UID]bool)
+				}
+				services = append(services, svcUID)
+				psc.podSvcUIDCache[pod.UID][svcUID] = true
+				psc.svcPodUIDCache[svcUID][pod.UID] = true
+			}
+		}
+	}
+	return services
+}
+
+// GetPodUIDsForService looks up a service in the cache and returns
+// the pods that match the services selector.
+func (psc *PodServiceCache) GetPodUIDsForService(svc *v1.Service) []types.UID {
+	return psc.GetPodUIDsForServiceUID(svc.UID)
+}
+
+// GetPodUIDsForServiceUID looks up a service in the cache and returns
+// the pods that match the services selector.
+func (psc *PodServiceCache) GetPodUIDsForServiceUID(svcUID types.UID) []types.UID {
+	var pods []types.UID
+	if podSet, exists := psc.svcPodUIDCache[svcUID]; exists {
+		for podUID := range podSet {
+			pods = append(pods, podUID)
+		}
+	}
+	return pods
+}
+
+// getServiceNamesForPod looks up a pod in the cache and returns its matching
+// services as a map in the form service_uid: service_name
+func (psc *PodServiceCache) getServiceNamesForPod(pod *v1.Pod) []string {
+	var services []string
+	if serviceUIDs, exists := psc.podSvcUIDCache[pod.UID]; exists {
+		for svcUID := range serviceUIDs {
+			services = append(services, psc.svcUIDNameCache[svcUID])
+		}
+	}
+	return services
+}
+
+// GetServiceNameForPod deterministically returns a single service
+// matched by a pods labels. Used by the DataPoint cache for
+// updating the "service" property, which only may have one value.
+// Selection method is sorting strings alphabetically and selecting the
+// first service off the top.
+func (psc *PodServiceCache) GetServiceNameForPod(pod *v1.Pod) (string, error) {
+	var service string
+	services := psc.getServiceNamesForPod(pod)
+	if len(services) == 0 {
+		return service, errors.New("no service matched pod")
```

Comment:
This is not very idiomatic for Go.  I would just make the return value for this func `*string` and then check for nil in the functions that call this to determine whether no service matched a pod or not.

Reply:
Errors are usually reserved for true errors, not things that are expected on a pretty regular basis.